vllm/model_executor/weight_utils.py

```diff
@@ -275,6 +276,122 @@ def hf_model_weights_iterator(
+    """
+
+    # TODO: Once a quantization params format is finalized, pull these Pydantic
+    # schemas out into a separate file that deals solely with quantization
+    # params and their related schemas so that they can be generalized and
+    # shared across various use cases.
+    class KVCacheQuantSchema(BaseModel):
+        dtype: str
+        # Each key is a TP rank. Each value is a dictionary mapping a TP rank's
+        # layer indices to their per-tensor KV cache scaling factor.
+        # TODO: Consider pulling this and its validation methods out into its
+        # own schema class (tricky as its members are variable)
+        scaling_factor: Dict[str, Dict[int, float]]
+
+        @computed_field
+        @property
+        def rank_keyword(self) -> str:
+            # Each TP rank key should be prefixed by a common rank_keyword.
+            # Thus, recovering the alphabetical components of any key should
+            # return it.
+            rank_keyword = "".join(
+                char for char in next(iter(self.scaling_factor.keys()))
+                if char.isalpha())
+            return rank_keyword
+
+        @model_validator(mode="after")
+        def check_is_fp8(self) -> "KVCacheQuantSchema":
+            assert self.dtype == "float8_e4m3fn", (
+                "Loaded scaling factors intended for KV cache dtype = "
+                f"{self.dtype} rather than float8_e4m3fn!")
+            return self
+
+        @model_validator(mode="after")
+        def check_tp_ranks(self) -> "KVCacheQuantSchema":
+            assert len(self.scaling_factor) == tp_size, (
+                f"Loaded dictionary has TP size {len(self.scaling_factor)} "
+                f"but LLM engine is currently running with TP size {tp_size}.")
+            for tp_rank, layer_maps in self.scaling_factor.items():
+                assert tp_rank.startswith(self.rank_keyword), (
+                    f"TP `{tp_rank}` does not start with `{self.rank_keyword}`"
+                )
+                assert len(layer_maps) == num_hidden_layers, (
+                    f"KV cache scales map for TP `{tp_rank}` is malformed. "
+                    f"Expected {num_hidden_layers} layers, got "
+                    f"{len(layer_maps)}.")
+            for i in range(tp_size):
+                assert f"{self.rank_keyword}{i}" in self.scaling_factor, (
+                    f"KV cache scales map for TP rank {i} not found.")
+            return self
+
+        @model_validator(mode="after")
+        def check_current_rank(self) -> "KVCacheQuantSchema":
+            layer_scales_map = self.scaling_factor[
+                f"{self.rank_keyword}{tp_rank}"]
+            for i in range(num_hidden_layers):
+                assert i in layer_scales_map, (
+                    f"Could not find KV cache scales for layer {i} in "
+                    f"TP rank {tp_rank}.")
+            return self
+
+    class QuantParamSchema(BaseModel):
```

Comment:
move this out of the function please. 

inline class definition has bad performance for evaluated each time.

Reply:
Performance should not be a significant consideration here. This is called at most tp_size times during the loading process and the runtime of this function is far eclipsed by the time it takes to load GB-scale weights. Good object-oriented decomposition considerations are more salient.